### PR TITLE
Fix the location of nametag on user video

### PR DESCRIPTION
**Pull request by Shaomei**
This PR takes the video size information from Zoom API config responses as well as onMyMediaChange event to make sure the nametag always display at the bottom right corner of user's video. It enables:
1) fixed location of nametag (not floating around when video is turned off/on);
2) relative sizing of the nametag to user's video;
3) relative resolution of the nametag, adaptive to user video feed resolution;

**After state:**
1. Low resolution camera (mirrored view)
Screenshot 2024-08-07 at 4 36 53 PM

2. High resolution camera (mirrored view)
Screenshot 2024-08-07 at 4 48 46 PM

### CODE REVIEW

File: lib/draw_badge_api.ts [hunk 1]
```diff
@@ -40,43 +41,121 @@ export class DrawBadgeApi {
...
-  }
+// TODO: make sure the handwave badge scale and resize correctly based on window size.
+function drawEverythingToImage(
+    media: Required<VideoMedia>,
+    nametag: NameTagBadge,
+    handWave: HandWaveBadge): ImageData {
+  const canvas = document.createElement('canvas');
+  const context = canvas.getContext('2d')!;
+  canvas.width = media.video.width!; // Width of the canvas
+  canvas.height = media.video.height!; // Height of the canvas
```
[the remark this answers is not in the document]
Comment (Shaomei) [reply lib/draw_badge_api.ts]: Done.

File: lib/draw_badge_api.ts [hunk 2]
```diff
@@ -40,43 +41,121 @@ export class DrawBadgeApi {
...
 
+  context.clearRect(0, 0, canvas.width, canvas.height);
+  if (nametag.visible) {
+    renderNameTagBadge(
+      nametag as EnabledNameTagBadge,
```
[the remark this answers is not in the document]
Comment (Shaomei) [reply lib/draw_badge_api.ts]: It works 🤓

File: lib/draw_badge_api.ts [hunk 3]
```diff
@@ -40,43 +41,121 @@ export class DrawBadgeApi {
...
+      canvas.width,
+      canvas.height,
```
[the remark this answers is not in the document]
Comment (Shaomei) [reply lib/draw_badge_api.ts:154]: Done

File: lib/zoomapi.ts [hunk 4]
```diff
@@ -1,8 +1,11 @@
-import zoomSdk, {ConfigOptions, ConfigResponse, GeneralMessageResponse }  from "@zoom/appssdk";
+import zoomSdk, {ConfigOptions, ConfigResponse, VideoMedia }  from "@zoom/appssdk";
+
+export interface DrawImageCallback {
+    (v: Required<VideoMedia>): ImageData;
```
[the remark this answers is not in the document]
Comment (Shaomei) [reply lib/zoomapi.ts]: Done.

File: lib/zoomapi.ts [hunk 5]
```diff
@@ -11,24 +14,37 @@ export function createFromConfig(options: ConfigOptions) {
...
   private configResponse: null|Promise<ConfigResponse> = null;
+  private drawImageCallback: null|DrawImageCallback = null;
   constructor(private configOptions: ConfigOptions) {}
 
   initialize() {
     if (this.configResponse == null) {
       this.configResponse = zoomSdk.config(this.configOptions);
+      zoomSdk.onMyMediaChange((event) => {
+        console.log(event);
+        if (
```
[the remark this answers is not in the document]
Comment (Shaomei) [reply lib/zoomapi.ts]: Kind of..

File: lib/zoomapi.ts [hunk 6]
```diff
@@ -11,24 +14,37 @@ export function createFromConfig(options: ConfigOptions) {
...
+    await this.drawForeground(configResponse.media);
   }
 
-  async removeVirtualForeground(): Promise<GeneralMessageResponse> {
-    await this.initialize();
-    return zoomSdk.removeVirtualForeground();
+  private async drawForeground(media: VideoMedia|undefined) {
+    if (this.drawImageCallback == null) return;
+    if (media?.video?.width == null) return;
+    const imageData = this.drawImageCallback(media as Required<VideoMedia>);
```
[the remark this answers is not in the document]
Comment (Shaomei) [reply lib/zoomapi.ts]: Done

**Review by blickly — APPROVED** [review]

**Pull request merged**